### ISSUE TITLE
ansible-lint  runs in offline mode by default

### ISSUE THREAD

**Issue by konstruktoid** [ansible/ansible-lint#3607]
##### Summary

`ansible-lint`  runs in `offline mode` by default without no option of disabling it.

##### Issue Type

- Bug Report

##### OS / ENVIRONMENT

```sh
$ cat .ansible-lint 
---
profile: production
exclude_paths:
  - .git/
  - .github/
$ ansible-lint --version
ansible-lint 6.17.3.dev7 using ansible-core:2.15.1 ansible-compat:4.1.2 ruamel-yaml:0.17.26 ruamel-yaml-clib:0.2.7
You are using a pre-release version of ansible-lint.
$ ansible-lint .
WARNING  Skipped installing collection dependencies due to running in offline mode.

Passed: 0 failure(s), 0 warning(s) on 13 files. Profile 'production' was required, but only 'production' profile passed.
You are using a pre-release version of ansible-lint.
```
`python3 -m pip install -U ansible git+https://github.com/ansible-community/ansible-lint.git`

##### STEPS TO REPRODUCE

See above.

##### Desired Behavior

Don't run in offline mode by default.

##### Actual Behavior

See above.

**Comment by konstruktoid** [comment]
GHA also affected: https://github.com/konstruktoid/ansible-cowrie-rootless/actions/runs/5601055483/job/15215642790#step:4:1

**Comment by rustydb** [comment]
I see it occurring in v6.17.2 and v6.18.0, always runs in offline mode.

**Comment by ssbarnea** [comment]
I am unable to reproduce this with current v6.20.0 release, I did `ansible-lint -vv ...|grep offline` and I see that offline mode is not enabled.

**Comment by konstruktoid** [comment]
```sh
$ cat .ansible-lint 
---
profile: production
exclude_paths:
  - .git/
  - .github/
$ ansible-lint --version
ansible-lint 6.20.0 using ansible-core:2.15.4 ansible-compat:4.1.10 ruamel-yaml:0.17.32 ruamel-yaml-clib:0.2.7
$ ansible-lint 
WARNING  Skipped installing collection dependencies due to running in offline mode.
Passed: 0 failure(s), 0 warning(s) on 33 files. Profile 'production' was required, and it passed.
$ ansible-lint -vv &> /tmp/log && grep -i offline /tmp/log 
DEBUG    Options: Options(cache_dir=PosixPath('/home/tsj/.cache/ansible-compat/ca6d44'), colored=True, configured=True, cwd=PosixPath('/home/tsj/Code/Git/ansible-role-docker-rootless'), display_relative_path=True, exclude_paths=['.git/', '.github/'], format=None, lintables=[], list_rules=False, list_tags=False, write_list=[], parseable=False, quiet=False, rulesdirs=[PosixPath('/home/tsj/Code/venvs/ansible/lib/python3.11/site-packages/ansiblelint/rules')], skip_list=[], tags=[], verbosity=2, warn_list=['experimental', 'jinja', 'fqcn'], mock_filters=[], mock_modules=[], mock_roles=[], loop_var_prefix=None, only_builtins_allow_collections=[], only_builtins_allow_modules=[], var_naming_pattern=None, offline=False, project_dir='.', extra_vars=None, enable_list=[], skip_action_validation=True, strict=False, rules={}, profile='production', task_name_prefix='{stem} | ', sarif_file=None, config_file='/home/tsj/Code/Git/ansible-role-docker-rootless/.ansible-lint', generate_ignore=False, rulesdir=[], use_default_rules=False, version=False, list_profiles=False, ignore_file=None, max_tasks=100, max_block_depth=20)
WARNING  Skipped installing collection dependencies due to running in offline mode.
```

**Comment by docsaistvan** [comment]
+1

**Comment by JCTechSol** [comment]
@ssbarnea I am still seeing this behavior in ansible-lint 6.20.3 & 6.21.1.

```
Options: Options(... offline=False....
...
WARNING  Skipped installing collection dependencies due to running in offline mode.
```

**Comment by danielkubat** [comment]
@ssbarnea same here:

code snippet of action I'm using:

```
steps:
  - name: Git checkout
    uses: actions/checkout@v4

  - name: Lint Ansible roles
    uses: ansible/ansible-lint@v6.21.1
```

result:

![image](https://github.com/ansible/ansible-lint/assets/42444343/49f199c1-3ab4-4e26-91d3-dc0f9bc4edba)

**Comment by szEvEz** [comment]
I am also running into this issue using version v6.22.0. Was there an older working version of this?

**Comment by ziegenberg** [comment]
Please re-open as this is still happening. There are multiple examples in the comments above.

**Comment by goldyfruit** [comment]
Same here.

```
WARNING  Skipped installing collection dependencies due to running in offline mode.

Passed: 0 failure(s), 0 warning(s) on 3 files. Last profile that met the validation criteria was 'production'.
```

**Comment by ademmers** [comment]
Note: this appears to be a warning only, as  far as I can see content is actually installed.

As far as I can trace it could be caused by https://github.com/ansible/ansible-lint/pull/3774

**Comment by reinout** [comment]
I did some "print" debugging in my install of `6.22.1` (pipx, osx). The #3774 mentioned by @ademmers seems to be the cause of the offline behaviour.

The default value for `_skip_ansible_syntax_check` in `config.py` is `False`.

In `runner.py`, https://github.com/ansible/ansible-lint/commit/0920931cae7b0fe3b0326c7fc1c83c80713ef016#diff-bcd284f890e371be042a5d19c8dbb4a27d497f2c5054746488173e5a17a3d117R239-R240, this code is run:

        if not self.skip_ansible_syntax_check:
            app = get_app(offline=True)

As `skip_ansible_syntax_check` is `False` by default, `not False` evaluates to `True`, so **by default the app is initialised in offline mode**!

(Note: I'm having serious trouble reasoning about a "triple negative": `if not` coupled with `skip_` and a default of `False`... a `_run_ansible_syntax_check` with a default of `True` would help, I think.)

**Comment by ssbarnea** [comment]
Have you searched for all calls to get_app? There are several and that one is offline on purpose to avoid performing some tasks again.

**Comment by reinout** [comment]
I didn't look at all calls of get_app, but I did put a `print` statement in the places with `skip_ansible_syntax_check`.
Maybe that led me to mis-diagnose it. I've now put a `print` in `get_app()`, right at the start and one after the `if default_options.offline != offline` couple of lines:

    $ ansible-lint ansible/*yml
    get_app, called with offline value None
    In the end, offline's value in options is None
    get_app, called with offline value True
    In the end, offline's value in options is True
    WARNING  Skipped installing collection dependencies due to running in offline mode.

    Passed: 0 failure(s), 0 warning(s) on 2 files. Last profile that met the validation criteria was 'production'.

So you're right that `get_app` is getting called (at least) twice. With the first time in "online" mode :-)

Two extra comments:

- In the mean time, I've fixed up my `requirements.yml` to include a minimum version number for `community.general`, as I was getting an error about `Unable to load module community.general.make`. *This* got ansible-lint to actually download a newer copy of the collection...
- I was confused by the highly visible "I'm running in offline mode" warning despite `community.general` being in my `requirements.yml`. Apparently the problem was not the offline mode, but a too-old `community.general` being cached somewhere.
